### PR TITLE
feat: implement relay service

### PR DESCRIPTION
**Pull request by amirvalhalla**
## Description

implement auto relay and relay service

### CODE REVIEW

**Review by kehiy — APPROVED** [review]

File: network/notifee.go [hunk 1]
```diff
@@ -84,6 +84,10 @@ func (s *NotifeeService) Stop() {
 	s.lp2pEventSub.Close()
 }
 
+func (s *NotifeeService) Reachability() lp2pnetwork.Reachability {
```
Comment (kehiy) [inline network/notifee.go:87]: also here you can consider RWlock to prevent this deadlock issue.

File: network/notifee.go [hunk 2]
```diff
@@ -84,6 +84,10 @@ func (s *NotifeeService) Stop() {
...
+	return s.reachability
```
Comment (kehiy) [inline network/notifee.go:88]: consider adding an RLock here to avoid the deadlock issue.

File: network/network.go [hunk 3]
```diff
@@ -259,6 +268,44 @@ func newNetwork(conf *Config, log *logger.SubLogger, opts []lp2p.Option) (*netwo
 	return n, nil
 }
 
+func findRelayPeers(h func() lp2phost.Host, log *logger.SubLogger) func(ctx context.Context,
+	num int) <-chan lp2ppeer.AddrInfo {
+	return func(ctx context.Context, num int) <-chan lp2ppeer.AddrInfo {
+		log.Debug("try to find relay peers", "num", num)
```
Comment (kehiy) [inline network/network.go:274]: I think it's better to use the `Info` level for logging here. what do you think?

**Review by kehiy — CHANGES_REQUESTED** [review]

File: network/notifee.go (hunk 2 above)
Comment (b00f) [reply network/notifee.go:88]: No need, This Will call only be the external API and only change once.

**Review by b00f — APPROVED** [review]

File: network/notifee.go (hunk 2 above)
Comment (b00f) [reply network/notifee.go:88]: The Readability is set only once, and the external API calls it. It is not necessary to add a Lock here. We have a test failure. Let's fix that.

Comment (kehiy) [reply network/notifee.go:88]: So, why there is a deadlock issue?

**Review by b00f — APPROVED** [review]

**Pull request merged**